diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6745e42..9e239f6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,6 +4,10 @@ project(libnfs
         LANGUAGES C
         VERSION 3.0.0)
 
+if(WINDOWS_STORE)
+  add_definitions(-D_WIN32_WINNT=0x0A00)
+endif()
+
 add_library(nfs
   lib/init.c
   lib/libnfs-sync.c
@@ -51,7 +55,7 @@ target_include_directories(
   $<INSTALL_INTERFACE:include/nfsc>
 )
 
-target_compile_definitions(nfs PRIVATE _U_=  _CRT_NONSTDC_NO_DEPRECATE _CRT_SECURE_NO_WARNINGS)
+target_compile_definitions(nfs PRIVATE _U_=  _CRT_NONSTDC_NO_DEPRECATE _CRT_SECURE_NO_WARNINGS _CRT_NONSTDC_NO_WARNINGS)
 target_link_libraries(nfs PRIVATE ws2_32.lib)
 set_target_properties(nfs PROPERTIES COMPILE_PDB_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR} COMPILE_PDB_NAME nfs)
 
-- 
2.21.0.windows.1

