diff --git a/libmariadb/CMakeLists.txt b/libmariadb/CMakeLists.txt
index 8cc86bc..3ee7aec 100644
--- a/libmariadb/CMakeLists.txt
+++ b/libmariadb/CMakeLists.txt
@@ -256,6 +256,7 @@ SET(MARIADB_NONBLOCK_SYMBOLS
 # handle static plugins
 SET(LIBMARIADB_SOURCES ${LIBMARIADB_PLUGIN_SOURCES})
 SET(SYSTEM_LIBS ${SYSTEM_LIBS} ${LIBMARIADB_PLUGIN_LIBS})
+INCLUDE_DIRECTORIES(BEFORE ${LIBMARIADB_PLUGIN_INCLUDES})
 ADD_DEFINITIONS(${LIBMARIADB_PLUGIN_DEFS})
 FOREACH(plugin ${PLUGINS_STATIC})
   SET(EXTERNAL_PLUGINS "${EXTERNAL_PLUGINS} extern struct st_mysql_client_plugin ${plugin}_client_plugin;\n")
diff --git a/plugins/io/CMakeLists.txt b/plugins/io/CMakeLists.txt
index 8c304c9..571b16d 100644
--- a/plugins/io/CMakeLists.txt
+++ b/plugins/io/CMakeLists.txt
@@ -1,7 +1,7 @@
 IF (WITH_CURL)
   INCLUDE(FindCURL)
   IF(CURL_FOUND)
-
+    INCLUDE_DIRECTORIES(BEFORE ${CURL_INCLUDE_DIR}) 
     ADD_DEFINITIONS(-DHAVE_REMOTEIO=1)
     #remote io plugin
     REGISTER_PLUGIN(TARGET remote_io
diff --git a/plugins/io/remote_io.c b/plugins/io/remote_io.c
index 1ac2d29..32bea50 100644
--- a/plugins/io/remote_io.c
+++ b/plugins/io/remote_io.c
@@ -92,7 +92,7 @@ typedef struct
 CURLM *multi_handle= NULL;
 
 #ifndef PLUGIN_DYNAMIC
-MARIADB_REMOTEIO_PLUGIN remote_io_plugin=
+MARIADB_REMOTEIO_PLUGIN remote_io_client_plugin =
 #else
 MARIADB_REMOTEIO_PLUGIN _mysql_client_plugin_declaration_ =
 #endif
-- 
2.16.1.windows.4

