diff --git a/ms/winrt.cpp b/ms/winrt.cpp
index efe31d8..e34bae4 100644
--- a/ms/winrt.cpp
+++ b/ms/winrt.cpp
@@ -74,6 +74,7 @@ int RAND_poll(void)
 	}
 
 #if defined(OPENSSL_WINAPP)
+#include<string>
 extern "C"
 {
 #include<stdio.h>
@@ -197,11 +198,39 @@ extern "C"
 					  const char *varname
 					  )
 		{
+		static std::string cert_path;
 		//hardcoded environmental variables used for the appx testing application for store/phone
 		if (!strcmp(varname, "OPENSSL_CONF"))
 			{
 			return "./openssl.cnf";
 			}
+		if (!strcmp(varname, "SSL_CERT_FILE"))
+		{
+		  if(cert_path.empty())
+		  {
+            auto values = Windows::Storage::ApplicationData::Current->LocalSettings->Values;
+            if (values->HasKey(L"SSL_CERT_FILE"))
+            {
+              auto value = safe_cast<Platform::String^>(values->Lookup(L"SSL_CERT_FILE"));
+              int len = WideCharToMultiByte(CP_UTF8, 0, value->Data(), -1, NULL, 0, NULL, NULL);
+              if (len == 0)
+                return "";
+
+              char *path = (char *)malloc(len);
+              if (!path)
+                return "";
+
+              if (WideCharToMultiByte(CP_UTF8, 0, value->Data(), -1, path, len, NULL, NULL) != len)
+              {
+                free(path);
+                return "";
+              }
+              cert_path = std::string(path);
+              free(path);
+            }
+		  }
+		  return (char*)cert_path.c_str();
+		}
 		return 0;
 		}
 	int winrt_setenv(const char *envname, const char *envval, int overwrite)
-- 
2.14.1.windows.1

