diff --git a/ms/winrt.cpp b/ms/winrt.cpp
index 4dcc8aa..dd754ba 100644
--- a/ms/winrt.cpp
+++ b/ms/winrt.cpp
@@ -87,6 +87,7 @@ int RAND_poll(void)
 	}
 
 #if defined(OPENSSL_WINAPP)
+#include<string>
 extern "C"
 {
 #include<stdio.h>
@@ -210,11 +211,24 @@ extern "C"
 					  const char *varname
 					  )
 		{
+		static std::string cert_path;
 		//hardcoded environmental variables used for the appx testing application for store/phone
 		if (!strcmp(varname, "OPENSSL_CONF"))
 			{
 			return "./openssl.cnf";
 			}
+		if (!strcmp(varname, "SSL_CERT_FILE"))
+		{
+			uint32_t varSize = GetEnvironmentVariableA(varname, nullptr, 0);
+			if (varSize <= 0)
+				return nullptr; // Not found
+
+			std::string value(varSize, 0);
+			GetEnvironmentVariableA(varname, const_cast<char*>(value.c_str()), varSize);
+
+			cert_path = value;
+			return const_cast<char*>(cert_path.c_str());
+		}
 		return 0;
 		}
 	int winrt_setenv(const char *envname, const char *envval, int overwrite)
-- 
2.14.1.windows.1

